hk2-locator/src/main/java/org/jvnet/hk2/internal/ServiceLocatorImpl.java

```diff
@@ -2224,6 +2224,12 @@ private void callAllConfigurationListeners(List<ServiceHandle<?>> allListeners)
 
     private Context<?> _resolveContext(final Class<? extends Annotation> scope) throws IllegalStateException {
         Context<?> retVal = null;
+        if (parent != null) {
+            retVal = parent._resolveContext(scope);
+            if (retVal != null) {
```

Comment:
So I stepped through this code.
Question: What happens when `parent != null` but it can't resolve `scope`? Is `! if (retVal != null)` (the opposite of current condition) even possible? It's all happening in this method, and it looks that if no `scope` is to be found the result is `IllegalStateException` being thrown.

:thinking:

Reply:
You’re right. I didn’t realise that the method throw exception instead of null. Would you want to raise a PR to catch the exception instead of checking null value?

Reply:
> the method throw exception instead of null. 

1. Now I think that perhaps the `parent` could be consulted only if the child can't find anything. Like the child could have something more appropriate, specialized. So use `parent` as the fallback. This would help a bit with p. 2 below.

> Would you want to raise a PR to catch the exception instead of checking null value?

2. That's private method. It can be modified to do anything. And the flow with catching exception instead of :question: (I have no solid idea at the moment) is the last thing I'd like to have. So at the moment I'm not ready to submit PR.

Reply:
It’s a private method, it means the only way it can be modified is to change its code. No way to change it in a subclass or anything else. If it changes, it should also adapt the code that handles the call to parent.

In fact, we can even change the method now to return null instead of exception, and wrap it in another method that throws an exception if the result is null after consulting both the parent and this service loader. That might be the cleanest option. Unless we want to reverse the order of parent and this as @buko suggested.